Lib/_pylong.py

```diff
@@ -83,6 +159,50 @@ def inner(n, w):
             result = -result
     return result
 
+old_int_to_decimal = int_to_decimal
+
+_unbounded_dec_context = decimal.getcontext().copy()
+_unbounded_dec_context.prec = decimal.MAX_PREC
+_unbounded_dec_context.Emax = decimal.MAX_EMAX
+_unbounded_dec_context.Emin = decimal.MIN_EMIN
+_unbounded_dec_context.traps[decimal.Inexact] = 1 # sanity check
+
+def new_int_to_decimal(n):
+    """Asymptotically fast conversion of an 'int' to Decimal."""
+
+    # Function due to Tim Peters.  See GH issue #90716 for details.
+    # https://github.com/python/cpython/issues/90716
+    #
+    # The implementation in longobject.c of base conversion algorithms
+    # between power-of-2 and non-power-of-2 bases are quadratic time.
+    # This function implements a divide-and-conquer algorithm that is
+    # faster for large numbers.  Builds an equal decimal.Decimal in a
+    # "clever" recursive way.  If we want a string representation, we
+    # apply str to _that_.
+
+    from decimal import Decimal as D
+    BITLIM = 200
+
+    def inner(n, w):
+        if w <= BITLIM:
+            return D(n)
+        w2 = w >> 1
+        hi = n >> w2
+        lo = n - (hi << w2)
```

Comment:
> If the goal is simplification of the code

The primary goal, yes, but not the only. We'll have to disagree about whether it's simpler. As I noted when reviewing your code, the caching scheme you made up, while brief, was quite unlike the caching scheme the other two functions used (which were _very_ similar to each other). That raises the bar for understanding the code as a whole. _All_ caching logic is one place now.

For the other two functions, their caching logic consumed a very large proportion of their total code.

`compute_powers()` is larger than any of the earlier schemes because it's trying to do the best (reasonably) possible job of computing all the needed powers quickly. That's _a_ goal too, but not the primary one.

Reply:
I don't know. From previous experience, I know that the relative speeds of CPython's bit-fiddling operations varies a whole lot across platforms. I picked the spelling that requires 2 bigint operations instead of the 3 in your alternative, minimizing the number of trips around the eval loop :wink:.

In this context, though, it doesn't matter - both are `O(n.bit_length())`, insignificant compared to the multiply on the next line.

Reply:
Trying it, your suggestion is significantly faster for very large ints. So I'll adopt it. Thanks! I got into the habit for code using small ints, where minimizing the number of trips around the eval loop can actually matter.